### ISSUE TITLE
Conceptual discussion about _alt and _name attributes

### ISSUE THREAD

**Issue by arvislacis** [skripteria/wn-snowflake-plugin#2]
First of all, this is not an issue but more like my ideas and comments about `_alt` and `_name` attributes in the long term.

Before I go into details - I wanted to thank you about the plugin idea and it's current functionality - although it's in it's early stages (Beta version) and maybe lot of similar things can be done with Static Pages and snippets but I personally see that it has a lot of potential, especially if some more complex field types or site building/editing options will be introduced. I also see some areas where the current code base can be improved, for example, code clean-up (and using PSR-2 standart for code style), enumerated field types, new field types etc. So wait for some PR requests from me. :)

Now back to the topic - in my opinion, `_alt`, `_name` and overall syntax of primary field + `_` + alternative name may cause some issues in the future:
1. For example, some users may call one SF field `my_image` and another `my_image_alt` (e.g., as alternative image for something)  - as a two separate, different fields. Same thing can happen also with `file` type field and `_name`. In other words, if SF is widely used in site then this kind of variable name collision can happen without users noticing it.
2. If new similar attributes will be introduced then they always will need to be documented. For example, if `link` type would have new field for specifying target attribute then it probably be `_target` or similar - but it would be hard for developers to guess this value and it would also increase a chance for creating name collisions (as specified in 1. point) especially for already existing installations and changes so  backward compatibility would be hard to maintain.

I am not against the current system but I can see these both problems in future... As a possible solution I can propose to use different syntax for specifying these attributes - either be it `-` or `.`, for example, `-alt`, `-name` or `.alt`, `.name`. I personally probably like dot notation more, I see that something like this was used before but you simplified that in https://github.com/skripteria/wn-snowflake-plugin/commit/191137fbca92623409d4b71d5a4bc59eb4754936 commit. IMHO, `.alt`, `.name` would also be more logical for developers as they would be similar to array or property notation used in programming languages... Also there could be some kind of validator/validation later which would not allow or at least inform devs if they are using `-` or `.` symbols in SF Key names so these symbols and the second part of the primary field name would always be reserved for the future functionality and there would be lower risk for backward compatibility breaking changes.

Anyway, this is only my idea and comments. It would be nice if you could give some feedback about this but it's not mandatory.

**Comment by skripteria** [comment]
Thanks a lot for your input, I highly appreciate it!.
So far I have worked on my own and didnt get a lot of qualified feedback on it, so this is very valuable.

You are perfectly right with your observation about the . notation ( e.g. image.alt) in the earlier version. I decided to change it because I wasnt happy about the usability aspect of it.

Basicallly I figured out 3 alternatives so far:
- split the path and alt attribute to 2 different snowflake keys. This would take comfort for the site editors since they cant edit all both within 1 screen.
- use the myimage.alt notation, this also means that myimage isnt a string anymore but an array. So you cant have myimge for the path and myimage.alt for the attribute, its then myimage.path and myimage.alt. This felt not natural from the developlers standpoint, so I dropped it.
- 3rd solution is myimage and myimage_alt like its now implemented. 
I agree that this isnt elegant and has the potential for accasional name collisions, but the people who have to deal with it are mostly developers so it should be an acceptable caveat.

I dont want to go back to .alt notation anymore, maybe some sort of "ugly" change like 2 underscores instead of 1 (myimage__alt) would also help to minimize this risk? 

Ideas are welcome.

**Comment by arvislacis** [comment]
Yes, I can agree with you and can see how the dot notation can cause other problems - especially new problems with variable as an array. This approach may be useful only if there would be lot of different attributes for each of the fields or if there would be some kind of flexible functionality which would allow the devs to define their own custom attributes...

I looked again in the source code and can say that maybe I overreacted a bit because I can see that `_alt` and `_name` are processed only after their primary/parent Snowflake key is found so it reduces some risks. Anyways, my biggest concert which was also the reason of this discussion is - **what if a lot of new attributes will be introduced in the future?** Currently there is no problems with `_al`, `_name`, although, in my opinion, `_name` is very generic and it can also be accidentally used in other places or collide not only with Snowflake internal system but maybe even with different Twig filters and other plugin variables etc.

Also one other problem which the current notation causes - it's a bit hard to find Snowflake keys in the source code with `_name` because the `_name` part is used also in lot of other libraries etc. So, for example, if theme developer wants to find or update only problematic `_name` attributes then it would be a bit hard to find them without specific (and limiting) search conditions.

Yes, using your suggested `__alt`, `__name` option would be better in my opinion and it would be also easier to find these keys in the theme files. One another solution maybe would be to use a bit specific attribute names, for example, `_sfalt`, `_sfname` although it looks like a more uglier solution than `__` and would cause additional overload.

So probably 2 underscored could be the best option in the long term but, well, that's only my comments and ideas, no need to change anything. This all discussion is only a thing to keep in mind for the future. :)
